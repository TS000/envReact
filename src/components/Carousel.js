import React from 'react'
import Carousel from './reusable/Carousel'

/**
 * @type {React.Component}
 *
 * @description Create a Slider/Carousel using modern react. It's up to you to add styles.
 * Sass is available, but feel free to use any styling solution you. CSS-in-JS, CSS, etc.
 * This component needs to be reusable and customizable. Multiple instances of this component
 * should be able to exist in the same view.
 *
 * The Slider should:
 * a. Allow for variable slide intervals, but default to 4 seconds
 * b. Should pause when a user is interacting with the component
 * c. The Slider should be able to take different types of slides. For example,
 * it could be a single image or a set of tiles. Reference Beatport.com for an example
 */

// dummy data
const slides1 = [
  {
    title: '1',
    image: 'https://f4.bcbits.com/img/a1012855547_16.jpg'
  },
  {
    title: '2',
    image: 'https://f4.bcbits.com/img/a0953099894_10.jpg'
  },
  {
    title: '3',
    image: 'https://f4.bcbits.com/img/a4002538098_10.jpg'
  },
  {
    title: '13',
    image: 'https://f4.bcbits.com/img/a1174936089_10.jpg'
  }
]

const slides2 = [
  {
    title: '4',
    image: 'https://f4.bcbits.com/img/a1807035745_10.jpg'
  },
  {
    title: '5',
    image: 'https://f4.bcbits.com/img/a1104600288_16.jpg'
  },
  {
    title: '6',
    image: 'https://f4.bcbits.com/img/a1006880036_16.jpg'
  },
  {
    title: '7',
    image: 'https://f4.bcbits.com/img/a0898739411_16.jpg'
  },
  {
    title: '8',
    image: 'https://f4.bcbits.com/img/a1655308465_16.jpg'
  },
  {
    title: '9',
    image: 'https://f4.bcbits.com/img/a3549988168_10.jpg'
  },
  {
    title: '10',
    image: 'https://f4.bcbits.com/img/a0140465996_10.jpg'
  },
  {
    title: '11',
    image: 'https://f4.bcbits.com/img/a0015365024_10.jpg'
  },
  {
    title: '18',
    image: 'https://f4.bcbits.com/img/a1174936089_10.jpg'
  }
]

const size1 = 900
const size2 = 250
const speed = 4000

export const Slider = () => {
  return (
    <div className='slider'>
      <Carousel slides={slides1} speed={speed} size={size1} multi={false}/>
      <Carousel slides={slides2} speed={false} size={size2} multi={true}/>
    </div>
  )
}
