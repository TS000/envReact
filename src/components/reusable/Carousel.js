import React, { useState, useEffect, useLayoutEffect } from 'react'
import styled from 'styled-components'
import useIncrementor from '../../hooks/useIncrementor.js'

// logic to decide how big to make the carousel and slides
const CarouselWrapper = styled.div`
  color: #eee;
  height: ${(props) => props.size}px;
  width: ${(props) => parseFloat(props.multi ? props.size * 4 : props.size)}px;
  background-color: #ccc;
  margin: 0 auto;
  position: relative;
  overflow: hidden;
  margin-top: 40px;
`

const SlideContainer = styled.div`
  position: absolute;
  transition: transform 350ms;
  height: 100%;
  display: flex;
`
const Slide = styled.div`
    width: ${(props) => props.size}px;
    height: ${(props) => props.size}px;
    display: flex;
    align-items: center;
    justify-content: center;
    position: relative;
    transition: opacity 350ms;
    background-image: url('${(props) => props.background}');
    background-size: contain;
`
const NavArrows = styled.div`
  position: absolute;
  width: 15px;
  height: 15px;
  padding: ${(props) => (props.multi ? '10px' : '30px')};
  box-sizing: border-box;
  border-top: ${(props) => (props.multi ? '5px' : '20px')} solid white;
  border-right: ${(props) => (props.multi ? '5px' : '20px')} solid black;
  cursor: pointer;
  display: none;
  transition: 6s;
  top: 0;
  bottom: 0;
  margin: auto 0;

  ${/* sc-selector */ CarouselWrapper}:hover & {
    display: block;
  }
`

const NavArrowLeft = styled(NavArrows)`
  transform: rotate(-135deg);
  left: 20px;
  z-index: 2;
`

const NavArrowRight = styled(NavArrows)`
  transform: rotate(45deg);
  right: 20px;
  z-index: 2;
`

const Carousel = ({ slides, speed, size, multi }) => {
  const [moveSlide, setMoveSlide] = useState(0)
  const [auto, setAuto] = useState(true)
  const [totalSlides] = useState(slides.length)
  const [multiMaybe, setMultiMaybe] = useState(1)
  const [speedTime, setSpeedTime] = useState(4000)

  // custom hook for moving the slides automatically
  const { offset } = useIncrementor({
    total: totalSlides,
    enabled: auto,
    speed: speedTime,
    multi: multi
  })

  useEffect(() => {
    setMoveSlide(offset)
  }, [offset])

  // DRY
  const multiOrNot = () => {
    if (multi) {
      setMultiMaybe(4)
    }
  }

  const changeSpeed = (foo) => {
    if (speed > 0) {
      setSpeedTime(foo)
    } else {
      setSpeedTime(4000)
    }
  }

  // initial render check for layout properties
  useLayoutEffect(() => {
    multiOrNot()
    changeSpeed(speed)
  }, [])

  // logic for moving the slides
  const move = (input) => {
    if (input === 1 && moveSlide < totalSlides - multiMaybe) {
      setMoveSlide(moveSlide + 1)
      return
    }
    if (input === -1 && moveSlide > 0) {
      setMoveSlide(moveSlide - 1)
      return
    }
    if (input === -1 && moveSlide === 0) {
      setMoveSlide(totalSlides - multiMaybe)
      return
    }
    if (input === 1 && moveSlide === totalSlides - multiMaybe) {
      setMoveSlide(0)
      return
    }

    // if nothing above triggers we start the carousel again
    setAuto(true)
  }

  return (
    <CarouselWrapper
      multi={multi}
      size={size}
      onMouseEnter={() => setAuto(false)}
      onMouseLeave={() => setAuto(true)}
    >
      <NavArrowLeft
        aria-label="previous slide"
        multi={multi}
        onClick={() => {
          move(-1)
        }}
      />

      <SlideContainer
        style={{
          transform: `translate3d(-${moveSlide * size}px,0,0)`
        }}
      >
        {slides.map((slide) => (
          <Slide size={size} background={slide.image} key={slide.title} />
        ))}
      </SlideContainer>
      <NavArrowRight
        aria-label="next slide"
        multi={multi}
        onClick={() => {
          move(1)
        }}
      />
    </CarouselWrapper>
  )
}

export default Carousel
