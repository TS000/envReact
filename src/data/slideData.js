const data = [
  {
    title: '4',
    image: 'https://f4.bcbits.com/img/a1807035745_10.jpg'
  },
  {
    title: '5',
    image: 'https://f4.bcbits.com/img/a1104600288_16.jpg'
  },
  {
    title: '6',
    image: 'https://f4.bcbits.com/img/a1006880036_16.jpg'
  },
  {
    title: '7',
    image: 'https://f4.bcbits.com/img/a0898739411_16.jpg'
  },
  {
    title: '8',
    image: 'https://f4.bcbits.com/img/a1655308465_16.jpg'
  },
  {
    title: '9',
    image: 'https://f4.bcbits.com/img/a3549988168_10.jpg'
  },
  {
    title: '10',
    image: 'https://f4.bcbits.com/img/a0140465996_10.jpg'
  },
  {
    title: '11',
    image: 'https://f4.bcbits.com/img/a0015365024_10.jpg'
  },
  {
    title: '18',
    image: 'https://f4.bcbits.com/img/a1174936089_10.jpg'
  }
]

export default data

