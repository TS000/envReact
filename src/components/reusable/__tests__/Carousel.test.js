import React from 'react'
import { shallow } from 'enzyme'
import Carousel from '../Carousel'
import data from '../../../data/slideData'

describe('Carousel', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<Carousel slides={data} />)
    expect(wrapper).toMatchSnapshot()
  })
})
