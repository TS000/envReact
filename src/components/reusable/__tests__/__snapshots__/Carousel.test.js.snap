// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel renders correctly 1`] = `ShallowWrapper {}`;
